tools/tpm2_duplicate.c

```diff
@@ -0,0 +1,423 @@
+    }
+
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    const struct option topts[] = {
+      { "auth-key",              required_argument, NULL, 'p'},
+      { "inner-wrapper-alg",     required_argument, NULL, 'g'},
+      { "duplicate-key-private", required_argument, NULL, 'r'},
+      { "input-key-file",        required_argument, NULL, 'i'},
+      { "output-key-file",       required_argument, NULL, 'o'},
+      { "output-enc-seed-file",  required_argument, NULL, 's'},
+      { "parent-key",            required_argument, NULL, 'C'},
+      { "context",               required_argument, NULL, 'c'},
+    };
+
+    *opts = tpm2_options_new("p:g:i:C:o:s:r:c:", ARRAY_LEN(topts), topts, on_option,
+                             NULL, 0);
+
+    return *opts != NULL;
+}
+
+/**
+ * Check all options and report as many errors as possible via LOG_ERR.
+ * @return
+ *  true on success, false on failure.
+ */
+static bool check_options(void) {
+
+    bool result = true;
+
+    /* Check for NULL alg & (keyin | keyout) */
+    if (ctx.flags.g == 0) {
+        LOG_ERR("Expected key type to be specified via \"-G\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.key_type != TPM2_ALG_NULL) {
+        if((ctx.flags.i == 0) && (ctx.flags.o == 0)) {
+            LOG_ERR("Expected in or out encryption key file \"-k/K\","
+                    " missing option.");
+            result = false;
+        }
+        if (ctx.flags.i && ctx.flags.o) {
+            LOG_ERR("Expected either in or out encryption key file \"-k/K\","
+                    " conflicting options.");
+            result = false;
+        }
+    }
+    else
+    {
+        if (ctx.flags.i || ctx.flags.o) {
+            LOG_ERR("Expected neither in nor out encryption key file \"-k/K\","
+                    " conflicting options.");
+            result = false;
+        }
+    }
+
+    if (ctx.flags.C == 0) {
+        LOG_ERR("Expected new parent object to be specified via \"-C\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.flags.c == 0) {
+        LOG_ERR("Expected object to be specified via \"-c\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.flags.s == 0) {
+        LOG_ERR("Expected encrypted seed out filename to be specified via \"-S\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.flags.r == 0) {
+        LOG_ERR("Expected private key out filename to be specified via \"-r\","
+                " missing option.");
+        result = false;
+    }
+
+    return result;
+}
+
+static bool load_object(ESYS_CONTEXT *ectx, const char *arg, tpm2_loaded_object * obj) {
+    bool result = true;
+    tpm2_object_load_rc rc = tpm2_util_object_load(ectx, arg, obj);
+
+    if (rc != olrc_error) {
+        if (!obj->tr_handle) {
+            result = tpm2_util_sys_handle_to_esys_handle(ectx, obj->handle, &obj->tr_handle);
```

Comment:
is this is a common case, shouldn't the call to tpm2_util_object_load do this?

Reply:
Maybe but I feel it's a bit out of scope for this PR

Reply:
Typically you would have any library changes in one commit, then the tool or dependency on the second commit. You would scope your dependent commits on the PR to address the scope.

Reply:
Can we be clear what you're asking for here? Perhaps belongs in a separate ticket?

Reply:
I don't think this change needs a separate ticket. Right now this is all one big commit and you expressed concerns about granularity. If you're worried about granularity, breaking this into separate commits would be the right thing. One commit for this change, then one commit for the new tool.

Typically patch granularity matters if we needed to revert that patch, however, that's unlikely to happen, I don't see it introducing a bug so big its best to revert and try again so you can just leave it all as one commit.

PRs should be related by patch dependencies, so having these 2 commits on one PR would be appropriate. However, PR granularity isn't always enforced. However, smaller PRs are easier to review and merge, so keeping only required changes grouped by PR is good prudence.

You don't need a ticket for every little thing, but tickets for tools or bugs are always good.

Reply:
So to be clear, make that change to tpm2_util_object_load(), make a commit, and then add your tool as a commit.

git rebase -i origin/master is your friend here. I would make the change to the library as a commit, then make a change to the tool and commit, then use git rebase -i origin/master to change the history.

Reply:
What change? Do you mean move this code into tpm2_util_object_load() ?

     if (rc != olrc_error) {
        if (!obj->tr_handle) {
            result = tpm2_util_sys_handle_to_esys_handle(ectx, obj->handle, &obj->tr_handle);

Because I think you will find that will break a lot of existing tools

Reply:
Looks like 37c28eae3ad6d does some wild things I missed. This existing API is quite convoluted.

Reply:
Ill fix this and then rebase your commit on top

Reply:
Fix the other issues I mentioned please

Reply:
I _think_ they are complete, please let me know if I'm wrong

Reply:
@GlovePuppet 
I have 2 commits on branch https://github.com/williamcroberts/tpm2.0-tools/tree/tpm2-duplicate
* lib: fix the object load code to make it simple: https://github.com/williamcroberts/tpm2.0-tools/commit/7c7b48cc63d74c9f952fd6daf4ad5606494fab1f
* your patch amended to work with it: https://github.com/williamcroberts/tpm2.0-tools/commit/779083433eeaa91c840c087a571d40f33b63153a

You can pull these local and add them into your PR, or if you would like I can merge https://github.com/williamcroberts/tpm2.0-tools/commit/7c7b48cc63d74c9f952fd6daf4ad5606494fab1f independently. Let me know how you would like to proceed.

Reply:
I would prefer to get this tool merged (assuming you think it's ready) and then fix the library issue, it's been dragging on and my manager is starting to give me the stink eye :)